### PR TITLE
Hyperopt

### PR DESCRIPTION
**Pull request by imerad**
This PR adds a script run_hyperopt.py which performs hyperparameter tuning and benchmarking for a given a algorithm on a given dataset.

In the process, a dataset module was added with an extraction function that adds a custom amount of corruption to the training set.

Reproducibility through random seeding was also fixed and tested within linlearn and when the hyperopt script is called.

Number of scalar products computed by each algorithm is tracked during training.

### CODE REVIEW

File: linlearn/estimator/ch.py [hunk 1]
```diff
@@ -191,6 +191,7 @@ def grad(inner_products, state):
                         gradient[j + 1, k] = holland_catoni_estimator(
                             deriv_samples_outer_prods, eps
                         )
+                return 0
```
Comment (stephanegaiffas) [inline linlearn/estimator/ch.py:194]: pourquoi on retourne 0 ici et partout ailleurs ?

File: linlearn/estimator/tmean.py [hunk 2]
```diff
@@ -82,8 +82,8 @@ def partial_deriv(j, inner_products, state):
                 # TODO: Try out different sorting mechanisms, since at some point the
                 #  sorting order won't change much...
                 for k in range(n_classes):
-                    # partial_derivative[k] = trimmed_mean(deriv_samples[:, k], n_samples, percentage)
-                    partial_derivative[k] = fast_trimmed_mean(deriv_samples[:, k], n_samples, percentage)
+                    partial_derivative[k] = trimmed_mean(deriv_samples[:, k], n_samples, percentage)
```
Comment (stephanegaiffas) [inline linlearn/estimator/tmean.py:85]: pourquoi on est revenu au trimmed_mean initial ?

File: linlearn/learner.py [hunk 3]
```diff
@@ -427,7 +433,6 @@ def _get_solver(self, X, y):
                 penalty,
                 self.max_iter,
                 self.tol,
-                self.random_state,
```
Comment (stephanegaiffas) [inline linlearn/learner.py:430]: on ne seed pas les solvers avec random_state ?

File: linlearn/solver/saga.py [hunk 4]
```diff
@@ -139,7 +137,7 @@ def cycle(
 
                         weights[j, k] = w_new[j, k]
 
-                return max_abs_delta, max_abs_weight
+                return max_abs_delta, max_abs_weight, n_samples
```
Comment (stephanegaiffas) [inline linlearn/solver/saga.py:140]: si ça retourne en plus le nombre de produits scalaires, il me semble que dans saga c'est 2*n_samples non ?

File: experiment.py [hunk 5]
```diff
@@ -0,0 +1,481 @@
...
+
+min_mom_block_size = 0.001
+default_mom_block_size = 0.07
+max_mom_block_size = 0.2
+
+min_tmean_percentage = 0.001
+default_tmean_percentage = 0.01
+max_tmean_percentage = 0.3
+
+min_C_reg = -1
```
Comment (stephanegaiffas) [inline experiment.py:35]: Hmmm -1 c'est pas assez petit, normalement on utilise une grille 10^-4 ... 10^4 (et attention que ça soit en base 10 dans hyperopt)

File: linlearn/estimator/ch.py (hunk 1 above)
Comment (imerad) [reply linlearn/estimator/ch.py:194]: C'est un compteur des produits scalaires calculés par la fonction, mais c'est peut être pas la manière idéale de le faire ...

File: linlearn/estimator/tmean.py (hunk 2 above)
Comment (imerad) [reply linlearn/estimator/tmean.py:85]: Pour une raison que j'ignore encore la fonction fast_trimmed_mean est devenue beaucoup plus lente dans cette version du code (malgré qu'elle était plus rapide dans les expériences faites à l'extérieur de linlearn)

File: linlearn/learner.py (hunk 3 above)
Comment (imerad) [reply linlearn/learner.py:430]: On utilise random_state pour seeder les rng de numpy et numba dans l'init de BaseLearner lorsque random_state est fourni (comme dans wildwood). J'ai testé ça dans un notebook et ça fonctionne, les résultats sont reproductibles

File: linlearn/solver/saga.py (hunk 4 above)
Comment (imerad) [reply linlearn/solver/saga.py:140]: Hmm, je viens de revoir le code et je pense qu'on en calcule un seul pour chaque échantillon aléatoire, donc n_samples au total

File: experiment.py (hunk 5 above)
Comment (imerad) [reply experiment.py:35]: Ok, je change vers np.log(1e-4) et np.log(1e4)
